formats/properties/commonMain/src/kotlinx/serialization/properties/Properties.kt

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("FunctionName", "DeprecatedCallableAddReplaceWith")
+
+package kotlinx.serialization.properties
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.modules.*
+
+/**
+ * Transforms a [Serializable] class' properties into a single flat [Map] which consists of
+ * string keys and primitive type values, and vice versa. Located in separated `kotlinx-serialization-properties` artifact.
+ *
+ * If the given class has non-primitive property `d` of arbitrary type `D`, `D` values are inserted
+ * into the same map; keys for such values are prefixed with string `d.`:
+ *
+ * ```
+ * @Serializable
+ * class Data(val property1: String)
+ *
+ * @Serializable
+ * class DataHolder(val data: Data, val property2: String)
+ *
+ * val map = Properties.store(DataHolder(Data("value1"), "value2"))
+ * // map contents will be the following:
+ * // property2 = value2
+ * // data.property1 = value1
+ * ```
+ *
+ * If the given class has a [List] property `l`, each value from the list
+ * would be prefixed with `l.N.`, where N is an index for a particular value.
+ * [Map] is treated as a [key,value,...] list.
+ *
+ * @param serializersModule A [SerializersModule] which should contain registered serializers
+ * for [Contextual] and [Polymorphic] serialization, if you have any.
+ */
+public sealed class Properties(
+    override val serializersModule: SerializersModule,
+    ctorMarker: Nothing?
+) : SerialFormat {
+
+    private inner class OutMapper : NamedValueEncoder() {
```

Comment:
[nit] I am a fan of separating implementation to a different file, so users will see **only** public API, but I do not insist

Reply:
Given that these implementations are 30-40 lines long, I do not think this is really a problem to leave them in the same file